CASSGO-72 Fix connection issue in Amazon Keyspaces

yomipq opened the pull request:
This PR fixes #1873 

In Amazon Keyspaces, system.local returns the localhost address and system.peers returns a host that contains the same hostID as localhost. This causes the connection issue in refreshRing(), where host address is overwritten with the localhost address and a host with the same hostID results in the "cannot find host" error.

This commit fixes the issue by making map from a slice of hosts. This approach is also used in `func (s *Session) init()` https://github.com/apache/cassandra-gocql-driver/blob/trunk/session.go#L272-L275

dkropachev commented on host_source.go:
```diff
@@ -752,9 +752,14 @@ func refreshRing(r *ringDescriber) error {
 		return err
 	}
 
+	hostMap := make(map[string]*HostInfo, len(hosts))
+	for _, host := range hosts {
+		hostMap[host.HostID()] = host
+	}
+
```
I don't think this solution is straighforward, it relays on the fact that `r.GetHosts()` provides hosts in the certain order, if that order changes it stop working.
It would be better to update `GetHosts` to ignore hosts that are already in the list

yomipq replied:
Thank you for reviewing. I will try to fix `GetHosts()` .

yomipq replied:
I fixed `GetHosts()` to ignore localHost if the same Host ID exists.
Could you please confirm?

dkropachev replied:
looks good, can you please add comment why this code is there, mentioning issue number and `Amazon Keyspaces`

yomipq replied:
Yes, I added the comment.

yomipq committed:
Fix connection issue in Amazon Keyspaces

In Amazon Keyspaces, system.local returns the localhost address and
system.peers returns a host that contains the same hostID as localhost.
This causes the connection issue in refreshRing(), where host address is
overwritten with the localhost address and a host with the same hostID
results in the "cannot find host" error.

This commit fixes GetHosts() so that it ignores localhost if the same
hsotID is included in peerHosts.

jameshartig commented:
This needs a test.

Also, this addresses one of the issues but doesn't address the issue with the control connection querying the local table and getting back the wrong information. See https://github.com/apache/cassandra-gocql-driver/issues/1873#issuecomment-2778506348 for more details.

I was going to suggest that we do what that comment suggests and call `refreshRing` from within `setupConn` but that's going to end up causing double hosts lookups on initialization and it will not respect `DisableInitialHostLookup` if that's set.

The problem with `setupConn` is that it'll end up overwriting the host information with incorrect info since the local table has wrong information in it. It sounds like other drivers just do a full ring refresh but I don't think we can do that in this case. Should we just be ignoring some information in `addOrUpdate`? @joao-r-reis thoughts?

jameshartig commented on host_source.go:
```diff
@@ -722,7 +722,21 @@ func (r *ringDescriber) GetHosts() ([]*HostInfo, string, error) {
+
 	var partitioner string
 	if len(hosts) > 0 {
 		partitioner = hosts[0].Partitioner()
```
I think this needs to instead look at `localHost` instead because I don't think the peers table contains partitioner information.

joao-r-reis commented:
> The problem with setupConn is that it'll end up overwriting the host information with incorrect info since the local table has wrong information in it. It sounds like other drivers just do a full ring refresh but I don't think we can do that in this case. Should we just be ignoring some information in addOrUpdate? @joao-r-reis thoughts?

Yeah the fix is to make sure we do a full ring refresh instead of just updating the local host so that we can overwrite the local host info with the info from system.peers. Doing a full refresh even without any overwriting would also mitigate the issue because the current behavior of every host turning into 127.0.0.1 eventually after enough reconnects would no longer exist.

joao-r-reis commented:
The problem is that `(d *refreshDebouncer) refreshNow()` will attempt to get the control connection and if it's not up it will try to reconnect it so calling `refreshNow()` inside the control connection itself might not be ideal? This needs some more exploration to be honest. The easy solution would be to call the full refresh method bypassing the debouncer but at that point might as well remove the `refreshNow()` method of the debouncer and just use it for topology events.

yomipq commented:
Thank you for comments. I'm afraid but I don't understand the details for now. I will look into your comments and the codes.

joao-r-reis commented:
Closing this for now. I mentioned in https://github.com/apache/cassandra-gocql-driver/issues/1873#issuecomment-3436533650 that 2.0.0 might have fixed this already so we need someone that can verify whether CASSGO-72 can be closed or needs to stay open.
